x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvPercentile.java

```diff
@@ -0,0 +1,391 @@
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.esql.planner.PlannerUtils;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.esql.core.type.DataType.UNSIGNED_LONG;
+
+public class MvPercentile extends EsqlScalarFunction {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(
+        Expression.class,
+        "MvPercentile",
+        MvPercentile::new
+    );
+
+    /**
+     * 2^52 is the smallest integer where it and all smaller integers can be represented exactly as double
+     */
+    private static final double MAX_SAFE_LONG_DOUBLE = Double.longBitsToDouble(0x4330000000000000L);
+
+    private final Expression field;
+    private final Expression percentile;
+
+    @FunctionInfo(
+        returnType = { "double", "integer", "long" },
+        description = "Converts a multivalued field into a single valued field containing "
+            + "the value at which a certain percentage of observed values occur.",
+        examples = @Example(file = "mv_percentile", tag = "example")
+    )
+    public MvPercentile(
+        Source source,
+        @Param(name = "number", type = { "double", "integer", "long" }, description = "Multivalue expression.") Expression field,
+        @Param(name = "percentile", type = { "double", "integer", "long" }) Expression percentile
+    ) {
+        super(source, List.of(field, percentile));
+        this.field = field;
+        this.percentile = percentile;
+    }
+
+    private MvPercentile(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(field);
+        out.writeNamedWriteable(percentile);
+    }
+
+    @Override
+    protected Expression.TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        return isType(field, dt -> dt.isNumeric() && dt != UNSIGNED_LONG, sourceText(), FIRST, "numeric except unsigned_long").and(
+            isType(percentile, dt -> dt.isNumeric() && dt != UNSIGNED_LONG, sourceText(), SECOND, "numeric except unsigned_long")
```

Comment:
Percentile seems to require a foldable `percentile` value. Just curious why mv_percentile can accept even fields.

Reply:
In the agg case, it's because it's "impossible" for it to not be foldable (Unless it appears in the grouping, which is a case I'm not sure we're allowing btw).
The evaluator however, can handle them without problem, so I think it's better to just allow it. Simpler function, less arbitrary restrictions

Reply:
I'm not saying it's wrong, but maybe suggest to look at this from a different angle: language consistency. If one function does it one way and the other does it differently, we should try to be consistent and have the same behavior; this makes it less confusing for users and easier to use/adopt.

Reply:
The consistent way for me would be to not force foldables on non-agg functions, as they all can be evaluated whether foldable or not. If I as a user tried to use a field in a function, and it told me that it must be a constant, _that_ would be confusing 👀 
For aggs, it's a logical problem, as they need X fields to accumulate, and Y fields as configuration.

So I'm really not seeing the problem of having non-foldable fields. I feel like we would be placing arbitrary restrictions

Reply:
Since `PERCENTILE` has to hit the whole index it has to be foldable. `MV_PERCENTILE` doesn't really have to be. But it's certainly kindest if, when it is, we can fail the query up front for invalid stuff.

I'm kind of ambivalent on forcing it to be foldable - I like that if we force it now we can always un-force it later. We can't go the other way. But in a lot of ways that's actually more work for a less flexible system.